stable_diffusion/stable_diffusion_tpu.py

```diff
@@ -0,0 +1,140 @@
+"""Ray Serve Stable Diffusion example."""
+from io import BytesIO  # pylint:disable=g-importing-member
+from typing import List
+from fastapi import FastAPI  # pylint:disable=g-importing-member
+from fastapi.responses import Response  # pylint:disable=g-importing-member
+import ray
+from ray import serve
+
+app = FastAPI()
+_MAX_BATCH_SIZE = 64
+
+@serve.deployment(num_replicas=1)
+@serve.ingress(app)
+class APIIngress:
+    def __init__(self, diffusion_model_handle) -> None:
+        self.handle = diffusion_model_handle
+
+    @app.get(
+        "/imagine",
+        responses={200: {"content": {"image/png": {}}}},
+        response_class=Response,
+    )
+    async def generate(self, prompt: str):
+        assert len(prompt), "prompt parameter cannot be empty"
+
+        image = await self.handle.generate.remote(prompt)
+        return image
+
+
+@serve.deployment(
+    ray_actor_options={
+        "resources": {"TPU": 4},
+    })
+class StableDiffusion:
+  """FLAX Stable Diffusion Ray Serve deployment running on TPUs.
+
+  Attributes:
+    run_with_profiler: Whether or not to run with the profiler. Note that
+      this saves the profile to the separate TPU VM.
+
+  """
+
+  def __init__(
+      self, run_with_profiler: bool = False, warmup: bool = False,
+      warmup_batch_size: int = _MAX_BATCH_SIZE):
+    from diffusers import FlaxStableDiffusionPipeline  # pylint:disable=g-import-not-at-top,g-importing-member
+    from flax.jax_utils import replicate  # pylint:disable=g-import-not-at-top,g-importing-member
+    import jax  # pylint:disable=g-import-not-at-top,unused-import
+    import jax.numpy as jnp  # pylint:disable=g-import-not-at-top
+    from jax import pmap  # pylint:disable=g-import-not-at-top,g-importing-member
+
+    model_id = "CompVis/stable-diffusion-v1-4"
+
+    self._pipeline, params = FlaxStableDiffusionPipeline.from_pretrained(
+        model_id,
+        revision="bf16",
+        dtype=jnp.bfloat16)
+
+    self._p_params = replicate(params)
+    self._p_generate = pmap(self._pipeline._generate)
+    self._run_with_profiler = run_with_profiler
+    self._profiler_dir = "/tmp/tensorboard"
+
+    if warmup:
+      print("Sending warmup requests.")
```

Comment:
Can you use [the Serve logger](https://docs.ray.io/en/latest/serve/monitoring.html#ray-logging) instead of `print`? The same comment applies to all the `print` statements across the file.

Reply:
Done in [b6c0923](https://github.com/ray-project/serve_config_examples/pull/8/commits/b6c092377f94e9486a07bdae81dc4e36d0b612b7).